python/spark_sklearn/group_apply.py

```diff
@@ -0,0 +1,158 @@
+"""
+Method for applying arbitrary UDFs over grouped data.
+"""
+
+from distutils.version import LooseVersion
+from itertools import chain
+from py4j import java_gateway
+
+from pyspark import SparkContext
+from pyspark.sql import DataFrame, SparkSession
+from pyspark.sql.functions import explode, udf
+from pyspark.sql.types import *
+from pyspark.sql.types import Row
+
+
+def gapply(grouped_data, func, schema, *cols):
+    """WARNING: this function is planned to be migrated to apache/spark for release 2.1. This
+    API will not be maintained.
+
+    Applies the function `func` to the grouped data; in particular, by default this calls
+    `func(key1, key2, ..., keyn, values)` where the number and order of the key arguments is
```

Comment:
I considered this at first (it's actually an easier impl). However, I have to disagree with the benefits of this:
1. It's arguable that the user *should* know the actual structure of the key. This is sort of like a type-check.
2. It's really inconvenient because most cases have a single key, and unwrapping that is annoying.
3. It's possible to avoid structure if you really don't want to by defining your function as `f(*args)`, then letting `args[-1]` be values.